# Estructuras y Formato de los Datos

## 1. Importamos las librerias

In [38]:
import pandas as pd
import numpy as np

# Para la visualización de datos
import matplotlib.pyplot as plt
from matplotlib import rcParams
#import seaborn as sns

# Estilo para las gráficas
plt.style.use("ggplot")
rcParams['figure.figsize'] = (12,6)

###  a. Leemos el archivo

In [39]:
df_processed=pd.read_csv('data/processed.cleveland.data',header= None)

df_processed.head()

0    1    2      3      4    5    6      7    8    9    10   11   12  13
0  63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0  0.0  2.3  3.0  0.0  6.0   0
1  67.0  1.0  4.0  160.0  286.0  0.0  2.0  108.0  1.0  1.5  2.0  3.0  3.0   2
2  67.0  1.0  4.0  120.0  229.0  0.0  2.0  129.0  1.0  2.6  2.0  2.0  7.0   1
3  37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0  0.0  3.5  3.0  0.0  3.0   0
4  41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0  0.0  1.4  1.0  0.0  3.0   0

#### Ponemos nombre a cada una de las columnas

In [40]:
# Nuestro dataset al cargarlo nos damos cuenta que no tienen nombre las columnas ,procedemos a colocarles nombre.
df_processed.columns= ['age','sex','cp','trestbps','chol','fbs', 'restecg','thalach','exang','oldepeak','slope','ca','thal','num']

df_processed.head()


age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldepeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0       2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0       1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0       2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0       3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0       1.4   

   slope   ca thal  num  
0    3.0  0.0  6.0    0  
1    2.0  3.0  3.0    2  
2    2.0  2.0  7.0    1  
3    3.0  0.0  3.0    0  
4    1.0  0.0  3.0    0

In [41]:
df_processed.thal.unique()

array(['6.0', '3.0', '7.0', '?'], dtype=object)

### b. Variables Numericas

In [42]:
# Análizamos el tipo de dato de cada variable. 
df_processed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldepeak  303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    object 
 12  thal      303 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


In [43]:
# Hacemos filtrado de columnas, 'ca', 'thal','oldepeak', para poder cambiar el tipo de dato a las variables numericas ya que estan son objeto.

df_processed_2=df_processed.drop(['ca', 'thal','oldepeak'], axis=1)
df_processed_2.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  slope  num
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0    3.0    0
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0    2.0    2
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0    2.0    1
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0    3.0    0
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0    1.0    0

#### Discretas:

Son valores que estan representados por números enteros

- Son: age, trestbps, chol, thalach

#### Continuas:

 Son valores estan representados por números decimales


- Son : oldepeak

### c. Variables Categoricas

Para poder análizar si las variables son númericas o Categoricas hemos cambiado el tipo de dato.

In [44]:
print(df_processed_2.astype(int))
# Cambiamos el tipo de número de float a entero ,para poder análizar bien los datos.

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  slope  num
0     63    1   1       145   233    1        2      150      0      3    0
1     67    1   4       160   286    0        2      108      1      2    2
2     67    1   4       120   229    0        2      129      1      2    1
3     37    1   3       130   250    0        0      187      0      3    0
4     41    0   2       130   204    0        2      172      0      1    0
..   ...  ...  ..       ...   ...  ...      ...      ...    ...    ...  ...
298   45    1   1       110   264    0        0      132      0      2    1
299   68    1   4       144   193    1        0      141      0      2    2
300   57    1   4       130   131    0        0      115      1      2    3
301   57    0   2       130   236    0        2      174      0      2    1
302   38    1   3       138   175    0        0      173      0      1    0

[303 rows x 11 columns]


#### Ordinal:

 Una variable puede ser tratada como ordinal cuando sus valores representan categorías con alguna clasificación intrínseca

- Son: ca, restcg 

#### Nominal:

 Cuando sus valores representan categorías 

- Son:  sex, fbs, cp, exang, slope, num, thal

In [45]:
df_ordenado=df_processed_2.astype(int)
df_ordenado=df_ordenado.sort_values(by=['age'], ascending=False)
df_ordenado

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  slope  num
161   77    1   4       125   304    0        2      162      1      1    4
257   76    0   3       140   197    0        1      116      0      2    0
233   74    0   2       120   269    0        2      121      1      1    0
103   71    0   3       110   265    1        2      130      0      1    0
42    71    0   2       160   302    0        0      162      0      1    0
..   ...  ...  ..       ...   ...  ...      ...      ...    ...    ...  ...
168   35    1   4       126   282    0        2      156      1      1    1
283   35    1   2       122   192    0        0      174      0      1    0
101   34    1   1       118   182    0        2      174      0      1    0
225   34    0   2       118   210    0        0      192      0      1    0
132   29    1   2       130   204    0        2      202      0      1    0

[303 rows x 11 columns]

### d. Análizamos valores faltantes


En este análisis encontramos que tenemos algunos valores faltantes en la columna de 'ca' y 'thal.  En este caso solo encontramos un solo tipo de signo  '?'. 

In [46]:
dt_dict= df_processed.thal.unique()
dt_dict

array(['6.0', '3.0', '7.0', '?'], dtype=object)

In [47]:
dt_dict=df_processed.ca.unique()
dt_dict

array(['0.0', '3.0', '2.0', '1.0', '?'], dtype=object)

In [48]:
df_processed[(df_processed["thal"] == "?") |  (df_processed["ca"] == "?")]


age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldepeak  \
87   53.0  0.0  3.0     128.0  216.0  0.0      2.0    115.0    0.0       0.0   
166  52.0  1.0  3.0     138.0  223.0  0.0      0.0    169.0    0.0       0.0   
192  43.0  1.0  4.0     132.0  247.0  1.0      2.0    143.0    1.0       0.1   
266  52.0  1.0  4.0     128.0  204.0  1.0      0.0    156.0    1.0       1.0   
287  58.0  1.0  2.0     125.0  220.0  0.0      0.0    144.0    0.0       0.4   
302  38.0  1.0  3.0     138.0  175.0  0.0      0.0    173.0    0.0       0.0   

     slope   ca thal  num  
87     1.0  0.0    ?    0  
166    1.0    ?  3.0    0  
192    2.0    ?  7.0    1  
266    2.0  0.0    ?    2  
287    2.0    ?  7.0    0  
302    1.0    ?  3.0    0

### e. Valores posibles que puede tomar la variable thal.

In [49]:
list(df_processed.thal.unique())

['6.0', '3.0', '7.0', '?']

In [50]:
df_processed[(df_processed["thal"]== "?" )]

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldepeak  \
87   53.0  0.0  3.0     128.0  216.0  0.0      2.0    115.0    0.0       0.0   
266  52.0  1.0  4.0     128.0  204.0  1.0      0.0    156.0    1.0       1.0   

     slope   ca thal  num  
87     1.0  0.0    ?    0  
266    2.0  0.0    ?    2

## Ejercicio 2

### a- Cargamos y leemos el dataFrame con el respectivo nombre de sus columnas.

In [51]:
dt_adul_data= pd.read_csv('data/adult.data.csv', header= None)
dt_adul_data.columns= ['age','workclass','fnlwgt','education','education-num','marital-status','occupation', 'relationship', 'race','sex','capital-gain','capital-loss','hours-per-week','native-country','income']
dt_adul_data.head(2)

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital-status        occupation    relationship    race    sex  \
0        Never-married      Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K

análizamos si tenemos valore faltantes y como estan representados

In [52]:
dt_adul_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [53]:
#Podemos observar que no existen valores nulos, pero sí  que contiene valores faltantes indeterminados presentados por '?'

columns_to_do_for = dt_adul_data.select_dtypes(include=['object']).columns
columns_that_have_missing_values = {}

for c in columns_to_do_for:
    columns_that_have_missing_values[c] = dt_adul_data[c].str.contains('?', regex=False).sum()

columns_that_have_missing_values

{'workclass': 1836,
 'education': 0,
 'marital-status': 0,
 'occupation': 1843,
 'relationship': 0,
 'race': 0,
 'sex': 0,
 'native-country': 583,
 'income': 0}

### b. Variables Númericas y Categóricas

#### Variables Categoricas

Ordinales: income ,educación

Nominal: sexo, race, education, marital_status, occupation, workclass, relationsship, native-country

#### Variables Numericas

Discretas : age, fnlwgt, education_num,  hours_week,

Continuas : hours_week, capital_loos, capital-gain

### c. Elementos repetidos

In [54]:
# Tenemos 32561 elementos en el dataset original.
total = dt_adul_data.shape
print(total)

(32561, 15)


In [55]:
# Tenemos 24 elementos repetidos o duplicados
duplicatec = dt_adul_data.duplicated().sum()
duplicatec

24

In [56]:
# Tenemos 32537 elementos unicos en el dataset original.
unicos = total[0] - duplicatec
print(unicos)

32537


### d. Creamos un nuevo datrame con valores duplicados

In [57]:
df_dup = dt_adul_data[dt_adul_data.duplicated()]
df_dup = df_dup.sort_values("age", ascending=False)

In [58]:
# ¿Cuántos de ellos tienen un país de origen indeterminado?
# Para los duplicados 1 personas tienen un país de origen indeterminado.  
df_dup['native-country'].str.contains('?', regex=False).sum()


1

In [59]:
# ¿Cuál es la proporción entre mujeres y hombres?  La proporción de mujeres a hombres es de 0.6 
df_dup.sex.value_counts(normalize=True)


sex
 Male      0.625
 Female    0.375
Name: proportion, dtype: float64

In [60]:
0.375/0.625

0.6

In [61]:
# ¿Cuántos cobran más de 50 mil dólares?

# Para los duplicados 2 personas ganan más de 50k
df_dup.income.value_counts()

income
 <=50K    22
 >50K      2
Name: count, dtype: int64

### e. Creamos un nuevo dataset con las entradas unicas

In [62]:
df_uni = dt_adul_data[dt_adul_data.duplicated()== False]

In [63]:
# ¿Cuántos tienen un país de origen indeterminado?
# Los que tienen un país de origen indeterminado son 582 
df_uni['native-country'].str.contains('?',regex=False).sum()

582

In [64]:
# ¿Cuántos tienen un tipo de trabajo indeterminado? 1836 entradas
df_uni.workclass.str.contains('?', regex=False).sum()

1836

In [65]:
# ¿Cuántos tienen ambas variables indeterminadas a la vez?
# Ambas variables tienen 27 personas.
ambos_indeterminados = df_uni[(df_uni['native-country'].str.contains('?', regex=False)) & (df_uni['workclass'].str.contains(pat='?',regex=False))]
ambos_indeterminados.shape[0]


27

In [66]:
# De entre quienes tienen un país de origen y un tipo de trabajo ambos indeterminados
# ¿cuál es la proporción entre hombres y mujeres? La proporción de hombre a mujer es de  0.799998560001152
# ¿Qué porcentaje de ellos cobra menos de 50k . Es de 14.8148%
porcentajes_sexo = ambos_indeterminados.sex.value_counts(normalize=True)
porcentajes_income = ambos_indeterminados.income.value_counts(normalize=True)

porcentajes_sexo, porcentajes_income 



(sex
  Female    0.555556
  Male      0.444444
 Name: proportion, dtype: float64,
 income
  <=50K    0.851852
  >50K     0.148148
 Name: proportion, dtype: float64)

In [67]:
0.444444/0.555556

0.799998560001152

### f. Busca en internet el significado de la variable “fnlwgt”. ¿Qué representa? Conociendo su significado, ¿crees que tiene sentido, desde un punto de vista de Machine Learning, eliminar los registros duplicados en este dataset?

Representa: En este caso la varible **fnlwgt** (peso final),  es el número de personas que el censo cree que representa la entrada. Entonces no tiene sentido tener una entrada con registros duplicados

### g. ¿Qué dificultades has encontrado a la hora de trabajar con los valores de las variables de este dataset? ¿Sería apropiado buscar una manera automática, usando métodos de Python, de eliminar esa dificultad o dificultades? ¿Se te ocurre cómo hacerlo? (En este apartado no hace falta   escribir   código,   basta   con   responder   en   texto   detallando   las ideas)

El encontrar valores nulos, puede resultar muy cansado siempre hacer una busqueda sobre valores faltantes o similiar

## Ejercicio 3

In [68]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [69]:
df_accident_bici_madrid_2019= pd.read_csv('data\AccidentesBicicletas_2019.csv',delimiter=';')
df_accident_bici_madrid_2019.head(2)

num_expediente       fecha      hora  \
0    2019S000036  02/01/2019  20:45:00   
1    2019S000045  03/01/2019  10:30:00   

                                      localizacion numero  cod_distrito  \
0  AVDA. GRAN VIA DE HORTALEZA / GTA. LUIS ROSALES    65B            16   
1                      CTRA. DEHESA DE LA VILLA, 1      1             9   

          distrito           tipo_accidente estado_meteorológico  \
0        HORTALEZA  Colisión fronto-lateral            Despejado   
1  MONCLOA-ARAVACA                    Caída            Despejado   

  tipo_vehículo tipo_persona       rango_edad    sexo  cod_lesividad  \
0     Bicicleta    Conductor  De 70 a 74 años  Hombre            3.0   
1     Bicicleta    Conductor  De 15 a 17 años  Hombre            2.0   

                        tipo_lesividad coordenada_x_utm coordenada_y_utm  \
0          Ingreso superior a 24 horas      444.591.127    4.479.784.701   
1  Ingreso inferior o igual a 24 horas      438.315.991    4.478.761.412   

  positiva_alcohol  positiva_droga  
0                N             NaN  
1                N             NaN

In [70]:
df_accident_bici_madrid_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   num_expediente        891 non-null    object 
 1   fecha                 891 non-null    object 
 2   hora                  891 non-null    object 
 3   localizacion          891 non-null    object 
 4   numero                891 non-null    object 
 5   cod_distrito          891 non-null    int64  
 6   distrito              891 non-null    object 
 7   tipo_accidente        891 non-null    object 
 8   estado_meteorológico  865 non-null    object 
 9   tipo_vehículo         891 non-null    object 
 10  tipo_persona          891 non-null    object 
 11  rango_edad            891 non-null    object 
 12  sexo                  891 non-null    object 
 13  cod_lesividad         795 non-null    float64
 14  tipo_lesividad        795 non-null    object 
 15  coordenada_x_utm      8

### a. Analizando el DataFrame, señala las variables numéricas y categóricas, especificando si son continuas,  discretas, ordinales o nominales.

#### Variables Numericas

Discretas:, , cod_districto, , rango_edad, fecha , cod_lesividad

Continuas: coordenada_y_utm,  coordenada_x_utm  , hora

#### Variables Categoricas

Ordinales: 

Nominales: sexo , localizacion, distrito, tipo_accidente, estado_meteorológico, tipo_vehículo, tipo_persona, tipo_lesividad, positiva_alcohol, positiva_droga, num_expediente, número

### b.¿Hay valores faltantes en el dataset? Especifica las variables que resentan valores faltantes y cuántos de ellos hay en cada variable

Valores faltantes, Nulos

In [71]:
# Si hay valores faltantes, en las variables:  positiva_droga 887 ,positiva_alcohol 2, positiva_alcohol 96,  cod_lesividad 96, estado_meteorológico 26  . y estan representados por un NaN.                    
df_accident_bici_madrid_2019.isna().sum()

num_expediente            0
fecha                     0
hora                      0
localizacion              0
numero                    0
cod_distrito              0
distrito                  0
tipo_accidente            0
estado_meteorológico     26
tipo_vehículo             0
tipo_persona              0
rango_edad                0
sexo                      0
cod_lesividad            96
tipo_lesividad           96
coordenada_x_utm          0
coordenada_y_utm          0
positiva_alcohol          2
positiva_droga          887
dtype: int64

### c. ¿Cuántos valores distintos toma la variable 'lesividad’?  Especifica esos valores. ¿Qué tipo de variable es la variable ‘lesividad’?

Se especifica en el pdf usar el archivo de 2019. En el cual hay dos columnas de lesividad

In [72]:
df_accident_bici_madrid_2019.tipo_lesividad.unique()

array(['Ingreso superior a 24 horas',
       'Ingreso inferior o igual a 24 horas',
       'Asistencia sanitaria sólo en el lugar del accidente',
       'Sin asistencia sanitaria',
       'Atención en urgencias sin posterior ingreso', nan,
       'Asistencia sanitaria inmediata en centro de salud o mutua',
       'Asistencia sanitaria ambulatoria con posterioridad',
       'Fallecido 24 horas'], dtype=object)

In [73]:
df_accident_bici_madrid_2019.cod_lesividad.unique()

array([ 3.,  2.,  7., 14.,  1., nan,  6.,  5.,  4.])

- Son 9  los tipos de valores distintos de la variable  tipo_lesividad, contando el NaN que representa un valor faltante.Es un tipo de variable Nominal.
- Sin 9 los valores distintos  de la variable cod_lesividad, contando el NaN que representa un valor faltante

### d. Ve al PDF de descripción de los datos, que se encuentra en el mismo enlace aportado en el enunciado. Busca el significado de los valore de lesividad. Explica   los   significados   para   cada   uno   de   los   valores posibles. 


LESIVIDAD:
01 Atención en urgencias sin posterior ingreso. - LEVE
02 Ingreso inferior o igual a 24 horas - LEVE
03 Ingreso superior a 24 horas. - GRAV
04 Fallecido 24 horas - FALLECIDO
05 Asistencia sanitaria ambulatoria con posterioridad - LEVE
06 Asistencia sanitaria inmediata en centro de salud o mutua - LEVE
07 Asistencia sanitaria sólo en el lugar del accidente - LEVE
14 Sin asistencia sanitaria
77 Se desconoce
En blanco Sin asistencia sanitaria.


##### Qué tipo de variable consideras que puede ser la variable lesividad?

Considero que es una variable Nominal

#### ¿Tu respuesta en este apartado es distinta a tu respuesta del apartado anterior?

Sigo considerando que es una variable  Nominal, ya que no cumple una jerarquía.

##### Cómo codificarías esta variable a la hora de utilizar este DataFrame como entrada de un modelo de Machine Learning?

De entrada para hacer un modelo de Machine Learnig utilizaría la variable del cod_lesividad en lugar del tipo_lesividad, porque solo acepta variable númericas un modelo de Machine Learning.

### e. Sin  tener   en   cuenta   los   valores   faltantes, ¿qué proporción de los accidentes registrados presentaban un test de alcohol positivo?

In [74]:
df_accident_bici_madrid_2019.positiva_alcohol.value_counts(normalize=True)

positiva_alcohol
N    0.987627
S    0.012373
Name: proportion, dtype: float64

Podemos comprobar que hay una proporción  de accidentes con positivo en alcohol  de 0.012373